executor/load_data.go

```diff
@@ -653,92 +612,41 @@ loop:
-			return nil, curData, false
-		}
-		var ok bool
-		curData, ok = e.getValidData(curData)
-		if !ok {
-			return nil, curData, false
+// ReadRows reads rows from parser. When parser's reader meet EOF, it will return
+// nil. For other errors it will return directly. When the rows batch is full it
+// will also return nil.
+// The result rows are saved in e.rows and update some members, caller can check
+// if curBatchCnt == 0 to know if reached EOF.
+func (e *LoadDataInfo) ReadRows(ctx context.Context, parser *mydump.CSVParser) error {
+	for e.IgnoreLines > 0 {
+		_, _, err := parser.ReadUntilTerminator()
+		if err != nil {
+			if errors.Cause(err) == io.EOF {
+				return nil
+			}
+			return err
 		}
-	}
-	var endIdx int
-	if ignore {
-		endIdx = strings.Index(string(hack.String(curData[startLen:])), e.LinesInfo.Terminated)
-	} else {
-		endIdx = e.indexOfTerminator(curData[startLen:])
-	}
-
-	if endIdx == -1 {
-		return nil, curData, true
-	}
 
-	return curData[startLen : startLen+endIdx], curData[startLen+endIdx+len(e.LinesInfo.Terminated):], true
-}
-
-// InsertData inserts data into specified table according to the specified format.
-// If it has the rest of data isn't completed the processing, then it returns without completed data.
-// If the number of inserted rows reaches the batchRows, then the second return value is true.
-// If prevData isn't nil and curData is nil, there are no other data to deal with and the isEOF is true.
-func (e *LoadDataInfo) InsertData(ctx context.Context, prevData, curData []byte) ([]byte, bool, error) {
-	if len(prevData) == 0 && len(curData) == 0 {
-		return nil, false, nil
+		e.IgnoreLines--
 	}
-	var line []byte
-	var isEOF, hasStarting, reachLimit bool
-	if len(prevData) > 0 && len(curData) == 0 {
-		isEOF = true
-		prevData, curData = curData, prevData
-	}
-	for len(curData) > 0 {
-		line, curData, hasStarting = e.getLine(prevData, curData, e.IgnoreLines > 0)
-		prevData = nil
-		// If it doesn't find the terminated symbol and this data isn't the last data,
-		// the data can't be inserted.
-		if line == nil && !isEOF {
-			break
-		}
-		// If doesn't find starting symbol, this data can't be inserted.
-		if !hasStarting {
-			if isEOF {
-				curData = nil
+	for {
+		if err := parser.ReadRow(); err != nil {
+			if errors.Cause(err) == io.EOF {
+				return nil
 			}
-			break
-		}
-		if line == nil && isEOF {
-			line = curData[len(e.LinesInfo.Starting):]
-			curData = nil
-		}
-
-		if e.IgnoreLines > 0 {
-			e.IgnoreLines--
-			continue
-		}
-		cols, err := e.getFieldsFromLine(line)
-		if err != nil {
-			return nil, false, err
+			return err
 		}
 		// rowCount will be used in fillRow(), last insert ID will be assigned according to the rowCount = 1.
 		// So should add first here.
 		e.rowCount++
-		e.rows = append(e.rows, e.colsToRow(ctx, cols))
+		e.rows = append(e.rows, e.colsToRow(ctx, parser.LastRow().Row))
 		e.curBatchCnt++
 		if e.maxRowsInBatch != 0 && e.rowCount%e.maxRowsInBatch == 0 {
```

Comment:
Why not `e.RowCount >= e.maxRowsInBatch`?

Reply:
`rowCount` will not reset, it's used as counter to report progress